A26-grpclb-selection.md

```diff
@@ -0,0 +1,135 @@
+A26: gRPCLB Selection
+----
+* Author(s): Doug Fawley
+* Approver: a11r
+* Status: Draft
+* Implemented in: None
+* Last updated: 2019-10-22
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+Modify the method for enabling gRPCLB to not require any special handling in
+core, allowing for stabilization of the gRPCLB protocol and load balancing
+policy.
+
+## Background
+
+The existing mechanism to enable and select gRPCLB is currently still
+experimental. In addition, gRPCLB itself is deprecated, to be replaced by xDS as
+the official advanced load balancing protocol for gRPC ([announcement
+email](https://groups.google.com/d/msg/grpc-io/0yGihF-EFQo/A4QKdXffBwAJ)). We
+would like to enable production uses of gRPCLB, but cannot do so if gRPCLB
+support is not stable. Therefore, we need a mechanism for selecting gRPCLB that
+can be declared stable, yet not create long-lasting maintenance problems.
+
+## Proposal
+
+C core will promote its existing channel arg that configures the DNS resolver to
```

Comment:
We _must_ have a stable API for client libraries to use it. I can agree that is separate from this proposal, but it was more a prerequisite.

I did not realize the "official word" is there is no stable API for users to control grpc.max_receive_message_length, grpc.max_connection_idle_ms, grpc.max_connection_age_ms, and similar in wrapped languages. I'm really confused why channel args were ever proposed as something Java should add to handle enabling SRV.

Reply:
All we really need for client libraries is to coordinate any changes with them ahead of time, which I'm fine with doing.  And while we've never promised stability for channel args strings, we also don't generally change them, so I don't expect this to cause any problem in practice.  I just don't want this gRFC to imply something to the world at large that we don't currently intend to imply.

As I said earlier, I do think that this is something that we should look more closely at in the future, but I think that's a broader topic, and I don't want to hold up this gRFC on that issue.  So, in the interest of moving forward with this gRFC, let's change the wording here to avoid the issue for now.

Reply:
> All we really need for client libraries is to coordinate any changes with them ahead of time

Are client libraries and gRPC expected to be kept in lockstep by users?  If not, CL should only be using stable gRPC APIs, or else users will run into problems when they attempt to use differing versions of the two libraries.

Reply:
Per their [documentation](https://github.com/googleapis/google-cloud-cpp#requirements), client libraries only specifies a *minimum* version of gRPC.  Any backward-incompatible changes to any gRPC APIs they use will cause breakages of their older versions.

Reply:
Oh, interesting.  I was under the impression that they built with a particular version linked directly into the client library.  But now that I think about it, I guess that would not work in many languages anyway.

In any case, I don't think this changes anything for this gRFC.  I believe (but @apolcyn can confirm) that the client libraries are already depending on this particular channel arg, and I'm comfortable with them doing so.  It's not a problem for us to ensure that we don't break that.  But I still don't want the language here to imply a broader promise.